src/H5VLint.c

```diff
@@ -705,6 +705,10 @@ H5VL_register(H5I_type_t type, void *object, H5VL_t *vol_connector, bool app_ref
         HGOTO_ERROR(H5E_VOL, H5E_CANTREGISTER, H5I_INVALID_HID, "unable to register handle");
 
 done:
+    if (ret_value < 0)
+        if (vol_obj && H5VL_free_object(vol_obj) < 0)
+            HDONE_ERROR(H5E_VOL, H5E_CANTRELEASE, H5I_INVALID_HID, "unable to release VOL object");
```

Comment:
This is required to properly clean up the VOL object if a failure occurs during `H5I_register()` even without the changes to `H5VL_t`'s nrefs field.